README.md
# chestnut

[![Gitter](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/plexus/chestnut?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)
[![Clojureverse](https://rawgit.com/clojureverse/clojureverse-assets/master/clojureverse-org-green.svg)](https://clojureverse.org/c/projects/chestnut)

[![Clojars Project](http://clojars.org/chestnut/lein-template/latest-version.svg)](http://clojars.org/chestnut/lein-template)

<img align="left" alt="Mr. Chestnut" src="resources/chestnut.png">


**Getting value out of Chestnut? Consider making [a small donation](https://patreon.com/plexus).**

Chestnut is a Clojure/ClojureScript application template. It takes the pain out
of getting a working ClojureScript setup with live reloading plus a browser
connected REPL.

It provides a solid default configuration for a REPL driven workflow, a
ClojureScript setup with separate dev/prod/test builds, Figwheel integration,
and a basic setup for running CLJ and CLJS unit tests. It provides a basic web
backend setup with Compojure, and a frontend based on one of the React wrappers
(your choice of Reagent, Rum, om-next, re-frame, or Vanilla JS). It works out of the
box on popular cloud providers like Heroku.

While Chestnut it's quite "complete" in a sense, it also tries to be minimal.
Boilerplate code is kept to a minimum, and there is only just enough structure
to get you started. How you build and structure your app from there is up to
you.

Chestnut should appeal to beginners because it allows you to get started quickly
and with minimal baggage, it should appeal to more seasoned developers because
it provides all the "obvious" configuration for a comfortable setup, without
being so opinionated that it becomes a straight jacket.

For deployment you get uberjar support, meaning you can get all your code
compiled, optimized, and packaged in a single executable JAR file. It also
contains the necessary artifacts to work on Heroku out of the box.

**Need help?** Ask [on the mailing list](http://clojureverse.org/c/chestnut)
(issues on GitHub are for bugs. Feature requests can be done either on the
mailing list or on Github.)

This README may describe unreleased features. Please compare the
[version number on Clojars](https://clojars.org/chestnut/lein-template) to the
[changelog below](https://github.com/plexus/chestnut/blob/master/README.md#changelog),
and check the README in your generated project for instructions pertaining to
your version.

## Usage

**This README may describe unreleased features, instead you can check the [README for the latest stable release](https://github.com/plexus/chestnut/tree/v0.15.1/README.md)**

```
lein new chestnut <name> <options>

e.g.

lein new chestnut my-app +garden +reagent +http-kit
```

If you're using the snapshot version, then make sure to add a `--` to separate Leiningen's and Chestnut's arguments

```
lein new chestnut my-app --snapshot -- +garden +reagent +http-kit
```

You can get an overview of all recognized options with

```
lein new chestnut +help
```

After that open the README of your generated project for detailed instructions.

### Lighttable

Lighttable provides a tighter integration for live coding with an inline
browser-tab. Rather than evaluating cljs on the command line with weasel repl,
evaluate code and preview pages inside Lighttable.

Steps: After running `(go)`, open a browser tab in Lighttable. Open a cljs file
from within a project, go to the end of an s-expression and hit Cmd-ENT.
Lighttable will ask you which client to connect. Click 'Connect a client' and
select 'Browser'. Browse to [http://localhost:10555](http://localhost:10555)

View LT's console to see a Chrome js console.

Hereafter, you can save a file and see changes or evaluate cljs code (without
saving a file). Note that running a weasel server is not required to evaluate
code in Lighttable.

### Emacs/Cider

Start a repl in the context of your project with `M-x cider-jack-in`.

After that it's the regular

```
(go)
(cljs-repl)
```

as described above.

## List of Contents

This template gives you everything you need to start developing
Clojure/ClojureScript apps effectively. It comes with

* [Figwheel](https://github.com/bhauman/lein-figwheel) Automatically reload your
  ClojureScript and CSS as soon as you save the file, no need for browser
  refresh.
* A ClojureScript interface to Facebook's React. You can choose between
  Reagent (default), Om-next (`+om-next`), re-frame (`+re-frame`),
  Rum (`+rum`), or use `+vanilla` to do without a React wrapper.
* [Ring](https://github.com/ring-clojure/ring) + Compojure. Clojure's de facto
  HTTP interface. Chestnut uses a Jetty or HttpKit server to serve the
  Clojurescript app. This way you already have an HTTP server running in case
  you want to add server-side functionality. Chestnut also inserts a Ring
  middleware to reload server-side Clojure files.
* Heroku support. Chestnut apps have all the bits and pieces to be deployable to
  Heroku. Getting your app on the web is as simple as `git push`.
* Unit tests for both Clojure and CLJS. Both specs and CLJS tests can be run in
  "auto" mode.

## Options

General options:

- `+help` Show an overview of all recognized options, then exit.
- `+no-poll` Opt out of usage statistics poll.
- `+http-kit` Use [HTTP Kit](http://http-kit.org/server.html) instead of Jetty.
- `+bidi` Use [bidi](https://github.com/juxt/bidi) instead of Compojure.
- `+site-middleware` Use the `ring.middleware.defaults.site-defaults` middleware (session, CSRF), instead of `ring.middleware.defaults.api-defaults` (see
  [ring.defaults documentation](https://github.com/ring-clojure/ring-defaults)).
- `+code-of-conduct` / `+coc` Add the [contributor covenant](http://contributor-covenant.org/) Code of Conduct.
- `+edge` Use the latest available version of all libraries/plugins. This includes alpha/beta versions, but does not include SNAPSHOT versions.
- `+bleeding-edge` Like `+edge`, but also use SNAPSHOT versions when available.

Choice of UI library:

- `+vanilla` Don't include Reagent, use this if you intend to use some other view library.
- `+om-next` Use om.next, instead of Reagent.
- `+re-frame` Use Reagent and re-frame.
- `+rum` Use Rum instead of Reagent.

Choice of CSS style:

- `+less` Use [less](https://github.com/montoux/lein-less) for compiling Less CSS files.
- `+sass` Use SASS for generating CSS.
- `+garden` Use Garden for generating CSS.

## Local copy

If you want to customize Chestnut, or try unreleased features, you can
run directly from master like this:

``` sh
git clone https://github.com/plexus/chestnut.git
cd chestnut
lein install
```

Note that master may be partially or wholly broken. I try to do
extensive manual testing before releasing a new stable version, so if
you don't like surprises then stick to the version on Clojars. Issue
reports and pull requests are very welcome.

## Requirements

* Java 1.7 or later
* Leiningen 2

## FAQ

* **Q:** How can I get the features in the SNAPSHOT version? <br>
  **A:** Use leiningen's `--snapshot` flag, e.g. `lein new chestnut
         my-project --snapshot`
* **Q:** I'm seeing warnings while compiling ClojureScript. <br>
  **A:** There are a few known warnings, but they should not affect the
         functioning of your app.
* **Q:** I changed the `{:text "Hello Chestnut!"}` portion and saved
         the file, but the changes don't show up. <br>
  **A:** It's a feature. The `app-state` is defined with `defonce`, so your
         application state doesn't reset every time you save a file. If you do
         want to reset after every change, change `(defonce app-state ..)` to
         `(def app-state ...)`.
* **Q:** I just want to compile ClojureScript to fully optimized JavaScript, so
         I can use it in a static HTML site. <br>
  **A:** Compile the "min" ClojureScript build, like this: `lein cljsbuild once min`, then look
         for `resources/public/js/app.js`.
* **Q:** I gave my project a very generic name like `cljs` or `clojure` and now
         it's not working. <br>
  **A:** This is due to namespace clashes. Try picking
         a more unique name. In particular avoid namespace prefixes used by
         Clojure, Clojurescript, or existing libraries.

## Sources

I used the
[browser-connected-repl](https://github.com/cemerick/austin/tree/master/browser-connected-repl-sample)
that's included with [Austin](https://github.com/cemerick/austin) as a
starting point, then pulled in bits from
[cljs-liveedit-webapp](https://github.com/ejlo/cljs-liveedit-webapp)
until things worked. Figwheel's
[Flappy Bird Demo app](https://github.com/bhauman/flappy-bird-demo)
also provided some ideas. The concept of refreshing Om when Figwheel
reloads was taken from
[this blog post](http://blog.michielborkent.nl/blog/2014/09/25/figwheel-keep-Om-turning/)
by [Michiel Borkent](https://github.com/borkdude).

For Heroku support I looked at Heroku's
[clojure-getting-started](https://github.com/heroku/clojure-getting-started)
example app.


## Other Templates

We hope Chestnut is great for you, but there are also many other templates that tackle
similar tasks.. Each has a slightly different perspective, so it is worth taking a look
at a few. Here is a [comparison chart](https://goo.gl/ZZH8fm) of many interesting
Clojure/ClojureScript templates.


## License

Copyright © 2014-2018 Arne Brasseur

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
## Changelog

### [0.19.0] (unreleased)

- Bring back `+reagent` / `--reagent` for compatibility

### [0.18.0] - 2019-04-03

- Update CIDER vars in dir-locals.el
- Bump Piggieback for latest jvm support

### [0.17.0] - 2019-01-01

- Drop Om, replace Reagent as the default UI library
- Update critically breaking dependencies
- Add javax xml bind dependency to fix Java 9+ compatability
- Explicitly add dotfiles to project for lein compatability
- Bump dependencies so project flags work out of the box

### [0.16.0] - 2018-05-06

- Add `+edge` and `+bleeding-edge` options for automatically using the latest version of dependencies.
- Upgrade ClojureScript, Om, Reagent, lein-cljsbuild, Figwheel
- Fix middleware order so static files correctly have their media type set
- Bring back logging to file, using [clj-logging-config](https://github.com/malcolmsparks/clj-logging-config)
- Add a help screen (`+help` or when using any unrecognized options)
- Set `goog.DEBUG` to `false` in the production ClojureScript build
- Fix lein 1.8 compatability
- Fix Procfile rendering

### [0.15.2] - 2017-07-30

- Update more docs to refer to `(go)` and `(cljs-repl)` instead of `(run)` and `(browser-repl)`
- Fix less and sass support
- Fix `index.html` generation when project name contains an underscore
- Upgrade several versions (ClojureScript, Ring, lein-cljsbuild, re-frame, piggieback, ring.middleware.logger)

### [0.15.1] - 2017-05-29

- Make the pop-poll work by shelling out to curl

### [0.15.0] - 2017-05-29

- Add Component support, both backend and frontend
- Add a `+re-frame` flag, for [re-frame](https://github.com/Day8/re-frame) projects
- remove `:repl-options {:init (go)}` from `project.clj`. You'll have to call `(go)` yourself after starting a REPL
- clean up the SASS/LESS support. It's still a hack, but a slightly cleaner hack
- Make `code_of_conduct.md` optional (`+coc` or `+code-of-conduct` flag to add it back)
- Make `browser-repl` and `run` deprecated, in favor of `cljs-repl` and `go`
- Collect usage statistics (version + command line flags). It is possible to opt-out with `+no-poll`.
- Bump versions: clojurescript 1.9.562, transit-clj 0.8.300, ring 1.6.1, ring-defaults 0.3.0, compojure 1.6.0, figwheel 0.5.10, figwheel-sidecar 0.5.10, org.clojure/tools.nrepl 0.2.13, om 1.0.0-alpha48

### 0.15.0-SNAPSHOT

- Options now use `+` as a prefix instead of `--`. This is more in line with
  other popular templates like Luminus, and it stops Leiningen from eating our
  options. The `--` variants still work.
- Add Rum support. Use (`+rum`).
- Add Om Next support. Use (`+om-next`).
- Bump versions: clojurescript 1.9.293, transit-clj 0.8.297, ring 1.5.0, ring-defaults 0.2.1, compojure 1.5.1, environ 1.1.0, reagent 0.6.0, figwheel 0.5.8, http-kit 2.2.0, om 1.0.0-alpha47, doo 0.1.7, lein-cljsbuild 1.1.5, lein-environ 1.1.0

### [0.14.0] - 2016-06-23

- Add cljc support
- ClojureScript 1.9.89
- Figwheel 0.5.4-4

### [0.13.0] - 2016-06-13

- Clojurescript 1.9.36
- Include Transit for faster ClojureScript compilation
- Figwheel 0.5.4-2
- Reagent 0.6.0-rc
- Om 1.0.0-alpha36
- Avoid using Leiningen profile merging for `:cljsbuild` configs

### [0.12.0] - 2016-05-24

* Upgrades
  * org.omcljs/om 1.0.0-alpha31 => 1.0.0-alpha35
  * lein-auto 0.1.1 => 0.1.2
  * lein-less 1.7.3 => 1.7.5
  * org.clojure/clojurescript 1.8.40 => 1.8.51
  * environ 1.0.2 => 1.0.3
  * lein-cljsbuild 1.1.1 => 1.1.3
  * lein-environ 1.0.1 => 1.0.3
  * figwheel 0.5.2 => 0.5.3-2
* Only add "dev" to the classpath in dev mode, fixes an issue with `lein jar`

### [0.11.0] - 2016-03-28

* Upgrade ClojureScript 1.7.228 => 1.8.40
* Upgrade Figwheel 0.5.1 => 0.5.2

### [0.10.1] - 2016-03-27

* Fix missing dependency when using `--http-kit`

### 0.10.0 - 2016-03-24

* Add switches for --reagent and --vanilla
* Upgrade Figwheel to 0.5.1
* Upgrade to Clojure 1.8.0, Om 1.1.0-alpha31, ring-defaults 0.2.0, compojure 1.5.0

### v0.9.1 - 2016-02-02

* Upgrade Figwheel to 0.5.0-6
* Upgrade Environ to 1.0.2
* Upgrade internal dependencies to play better with Boot

### v0.9.0 - 2016-01-29

* Use Figwheel for REPL and Ring server
* Remove boilerplate that's no longer needed
* Drop Enlive and simplify the dev setup
* Drop CLJX support
* Better inline documentation
* Upgrade all dependencies

### v0.8.2-SNAPSHOT
* Version upgrades: clojurescript: 1.7.145 => 1.7.189, om: 1.0.0-alpha15 => 1.0.0-alpha28, figwheel => 0.4.1, figwheel-sidecar 0.4.1 => 0.5.0-2, lein-figwheel 0.4.1 =>  0.5.0-2
* Version upgrades: clojurescript: 1.7.122 => 1.7.145, om: 0.9.0 => 1.0.0-alpha15, tools.nrepl => 0.2.12, figwheel => 0.4.1, figwheel-sidecar => 0.4.1, lein-figwheel => 0.4.1,environ => 1.0.1, lein-environ => 1.0.1
* Fix allowing projects to have - in their names.
* Fix allowing groupID/artifactID naming convention.
* Fix "No such namespace" when using om via the REPL.
* Fix specljs auto testing.
* Added gzip compression using [ring.middleware.gzip](https://github.com/bertrandk/ring-gzip).
* Added browser cache-control hinting using [ring.middleware.browser-caching](https://github.com/slester/browser-caching).
* Version upgrades: clojure: 1.6.0 => 1.7.0, clojurescript: 0.0-3058 => 1.7.48, ring: 1.3.2 => 1.4.0, ring-defaults: 0.1.4 => 0.1.5, compojure: 1.3.2 => 1.4.0, om: 0.8.8 => 0.9.0, figwheel: 0.2.5 => 0.3.7, piggieback: 0.1.5 => 0.2.1, weasel: 0.6.0 => 0.7.0

### v0.8.1

* Restore working lein repl for cljx projects ([Josh Daghlian](https://github.com/jcdcodes))
* Use latest version of enlive to fix windows issue ([Justin Squirek](https://github.com/AdaDoom3))

### v0.8.0

* Update Om: 0.8.0-rc1 => 0.8.8 ([Anna Pawlicka](https://github.com/annapawlicka))
* Make Uberjar executable ([Børge Svingen](https://github.com/bsvingen))
* Remove react.js injection in dev html ([Jamie English](https://github.com/english))
* Fix how Less and Sassc are invoked ([Daryl Lau](https://github.com/dlau))

src/leiningen/new/chestnut.clj
(ns leiningen.new.chestnut
  (:require [chestnut.httpclient :as http]
            [clj-jgit.porcelain :refer :all]
            [clojure.java.io :as io]
            [clojure.string :as str]
            [leiningen.core.main :as main]
            [leiningen.new.templates :refer [->files name-to-path project-name
                                             render-text renderer sanitize-ns slurp-resource]]
            [ancient-clj.core :as ancient])
  (:import java.io.Writer))

(def default-project-deps '[[org.clojure/clojure "1.10.0"]
                            [org.clojure/clojurescript "1.10.439" :scope "provided"]
                            [com.cognitect/transit-clj "0.8.313"]
                            [ring "1.7.1"]
                            [ring/ring-defaults "0.3.2"]
                            [bk/ring-gzip "0.3.0"]
                            [radicalzephyr/ring.middleware.logger "0.6.0"]
                            [clj-logging-config "1.9.12"]
                            [environ "1.1.0"]
                            [com.stuartsierra/component "0.3.2"]
                            [org.danielsz/system "0.4.1"]
                            [org.clojure/tools.namespace "0.2.11"]])

(def optional-project-deps '{:garden [lambdaisland/garden-watcher "0.3.3"]
                             :http-kit [http-kit "2.3.0"]
                             :compojure [compojure "1.6.1"]
                             :bidi [bidi "2.1.5"]
                             :lein-auto [lein-auto "0.1.3"]
                             :lein-less [lein-less "1.7.5"]
                             :lein-sassc [lein-sassc "0.10.5"]
                             :om-next [org.omcljs/om "1.0.0-beta4"]
                             :re-frame [re-frame "0.10.6"]
                             :reagent [reagent "0.8.1"]
                             :rum [rum "0.11.3"]})

(def default-project-plugins '[[lein-cljsbuild "1.1.7"]
                               [lein-environ "1.1.0"]])

(def project-clj-dev-deps '[[figwheel "0.5.18"]
                            [figwheel-sidecar "0.5.18"]
                            [cider/piggieback "0.4.0"]
                            [cider/cider-nrepl "0.18.0"]
                            [lein-doo "0.1.11"]
                            [reloaded.repl "0.2.4"]])

(def project-clj-dev-plugins '[[lein-figwheel "0.5.18"]
                               [lein-doo "0.1.11"]])

;; When using `pr`, output quoted forms as 'foo, and not as (quote foo)
(defmethod clojure.core/print-method clojure.lang.ISeq [o ^Writer w]
  (#'clojure.core/print-meta o w)
  (if (= (first o) 'quote)
    (do
      (.write w "'")
      (print-method (second o) w))
    (#'clojure.core/print-sequential "(" #'clojure.core/pr-on " " ")" o w)))

(def render (renderer "chestnut"))

(defn wrap-indent [wrap n list]
  (fn []
    (->> list
         (map #(str "\n" (apply str (repeat n " ")) (wrap %)))
         (str/join ""))))

(defn dep-list [n list]
  (wrap-indent #(str "[" % "]") n list))

(defn indent [n list]
  (wrap-indent identity n list))

(defn indent-next [n list]
  (str (first list)
       ((indent n (next list)))))

(def options-descriptions
  {"help"            "Show this help information and exit."
   "http-kit"        "Use the http-kit web server, instead of Jetty."
   "bidi"            "Use bidi for server side routing library instead of compojure."
   "site-middleware" "Use Ring's site-middleware, instead of the api-middleware. (session support)"
   "less"            "Use the LESS CSS pre-processor."
   "sass"            "Use the SASS CSS pre-processor."
   "vanilla"         "Don't use a UI library (vanilla Javascript)"
   "garden"          "Use Garden for writing CSS in Clojure"
   "rum"             "Use Rum as UI library"
   "om-next"         "Use om-next as UI library"
   "reagent"         "Use Reagent as UI library. (default)"
   "re-frame"        "Use re-frame as UI framework"
   "code-of-conduct" "Add a Code of Conduct to the project (Contributor Covenant)"
   "coc"             "Add a Code of Conduct to the project (Contributor Covenant)"
   "no-poll"         "Don't submit usage information"
   "edge"            "Automatically upgrade all libraries to the latest non-SNAPSHOT version."
   "bleeding-edge"   "Automatically upgrade all libraries to the latest non versions, inclusing SNAPSHOT releases."})

(def all-options
  ["http-kit"
   "bidi"
   "site-middleware"
   "less"
   "sass"
   "garden"
   "reagent"
   "re-frame"
   "rum"
   "om-next"
   "vanilla"
   "code-of-conduct"
   "coc"
   "edge"
   "bleeding-edge"
   "no-poll"
   "help"])

(doseq [opt all-options]
  (eval
   `(defn ~(symbol (str opt "?")) [opts#]
      (some #{~(str "--" opt) ~(str "+" opt)} opts#))))

(defn compojure? [opts]
  (not (bidi? opts)))

(defn reagent? [opts]
  (and (not (rum? opts))
       (not (vanilla? opts))
       (not (om-next? opts))
       (not (re-frame? opts))))

(defn some-edge? [opts]
  (or (edge? opts) (bleeding-edge? opts)))

(defn server-clj-requires [opts]
  (if (http-kit? opts)
    ["system.components.http-kit :refer [new-web-server]"]
    ["system.components.jetty :refer [new-web-server]"]))

(defn user-clj-requires [name opts]
  (cond-> []
    (or (sass? opts) (less? opts))
    (conj [(symbol (str (sanitize-ns name) ".components.shell-component :refer [shell-component]"))])

    (garden? opts)
    (conj '[garden-watcher.core :refer [new-garden-watcher]])))

(defn update-dep [dep opts]
  (cond-> dep
    (some-edge? opts)
    (assoc 1 (ancient/latest-version-string! dep {:snapshots? (bleeding-edge? opts)}))))

(defn update-deps [deps opts]
  (map #(update-dep % opts) deps))

(defn project-clj-deps [opts]
  (update-deps
   (cond-> default-project-deps
     (http-kit? opts) (conj (get optional-project-deps :http-kit))
     (compojure? opts) (conj (get optional-project-deps :compojure))
     (bidi? opts)     (conj (get optional-project-deps :bidi))
     (reagent? opts)  (conj (get optional-project-deps :reagent))
     (om-next? opts)  (conj (get optional-project-deps :om-next))
     (rum? opts)      (conj (get optional-project-deps :rum))
     (re-frame? opts) (conj (get optional-project-deps :re-frame))
     (garden? opts)   (conj (get optional-project-deps :garden)))
   opts))

(defn project-plugins [opts]
  (update-deps
   (cond-> default-project-plugins
     (sass? opts) (conj (get optional-project-deps :lein-sassc)
                        (get optional-project-deps :lein-auto))
     (less? opts) (conj (get optional-project-deps :lein-less)))
   opts))

(defn project-prep-tasks [name opts]
  (cond-> ["compile" ["cljsbuild" "once" "min"]]
    (garden? opts) (conj ["run" "-m" "garden-watcher.main" (str (sanitize-ns name) ".styles")])))

(defn project-uberjar-hooks [opts]
  (cond-> []
    (less? opts) (conj "leiningen.less")
    (sass? opts) (conj "leiningen.sassc")))

(defn extra-dev-components [name opts]
  (cond-> []
    (garden? opts)
    (into [:garden-watcher (list 'new-garden-watcher [(list 'quote (symbol (str (sanitize-ns name) ".styles")))])])

    (less? opts)
    (into [:less (list 'shell-component "lein" "less" "auto")])

    (sass? opts)
    (into [:sass (list 'shell-component "lein" "auto" "sassc" "once")])))

(defn load-props [file-name]
  (with-open [^java.io.Reader reader (clojure.java.io/reader file-name)]
    (let [props (java.util.Properties.)]
      (.load props reader)
      (into {} (for [[k v] props] [(keyword k) v])))))

(defn chestnut-version []
  (let [resource (io/resource "META-INF/maven/chestnut/lein-template/pom.properties")
        props (load-props resource)
        version (:version props)
        revision (:revision props)
        snapshot? (re-find #"SNAPSHOT" version)]
    (str version " (" (str/join (take 8 revision)) ")")))

(defn template-data [name opts]
  {:full-name            name
   :name                 (project-name name)
   :chestnut-version     (chestnut-version)
   :project-ns           (sanitize-ns name)
   :ns-name              (sanitize-ns name) ;; used by re-frame template
   :sanitized            (name-to-path name)

   :user-clj-requires    (indent 12 (map pr-str (user-clj-requires name opts)))
   :server-clj-requires  (dep-list 12 (server-clj-requires opts))

   :project-clj-deps     (indent-next 17 (map pr-str (project-clj-deps opts)))
   :project-plugins      (indent-next 12 (map pr-str (project-plugins opts)))
   :project-clj-dev-deps (indent-next 29 (map pr-str (update-deps project-clj-dev-deps opts)))
   :project-clj-dev-plugins (indent-next 24 (map pr-str (update-deps project-clj-dev-plugins opts)))
   :project-prep-tasks   (indent-next 27 (map pr-str (project-prep-tasks name opts)))
   :project-uberjar-hooks (str/join " " (project-uberjar-hooks opts))

   :server-command       (if (http-kit? opts) "run-server" "run-jetty")
   :ring-defaults        (if (site-middleware? opts) "site-defaults" "api-defaults")

   ;; TODO: get rid of these, instead use something like :project-clj-extras
   :less?                (less? opts)
   :sass?                (sass? opts)

   :compojure?           (compojure? opts)
   :bidi?                (bidi? opts)

   :extra-dev-components (indent 4 (->> (extra-dev-components name opts)
                                        (map pr-str)
                                        (partition 2)
                                        (map #(str/join " " %))))})

(defn files-to-render [opts]
  (cond-> ["project.clj"
           "resources/public/index.html"
           "resources/public/favicon.ico"
           "resources/log4j.properties"
           "src/clj/chestnut/application.clj"
           "src/clj/chestnut/routes.clj"
           "src/clj/chestnut/config.clj"
           "src/clj/chestnut/components/server_info.clj"
           "src/cljc/chestnut/common.cljc"
           "src/cljs/chestnut/system.cljs"
           "src/cljs/chestnut/components/ui.cljs"
           "dev/user.clj"
           "dev/cljs/user.cljs"
           "LICENSE"
           "README.md"
           "Procfile"
           ".gitignore"
           "system.properties"
           ".dir-locals.el"
           "test/clj/chestnut/example_test.clj"
           "test/cljs/chestnut/core_test.cljs"
           "test/cljs/chestnut/test_runner.cljs"
           "test/cljc/chestnut/common_test.cljc"]
    (or (coc? opts) (code-of-conduct? opts)) (conj "code_of_conduct.md")
    (less? opts) (conj "src/less/style.less")
    (sass? opts) (conj "src/scss/style.scss")
    (or (less? opts) (sass? opts)) (conj "src/clj/chestnut/components/shell_component.clj")
    (garden? opts) (conj "src/clj/chestnut/styles.clj")
    (not (or (less? opts) (sass? opts))) (conj "resources/public/css/style.css")))

(defn format-files-args
  "Returns a list of pairs (vectors). The first element is the file name to
  render, the second is the file contents."
  [{:keys [name] :as data} opts]
  (letfn [(render-file [file]
            [(str/replace file "chestnut" "{{sanitized}}")
             (render file data)])]
    (conj
     (map render-file (files-to-render opts))
     ["src/cljs/{{sanitized}}/core.cljs"
      (render (cond
                (reagent? opts) "src/cljs/chestnut/core_reagent.cljs"
                (om-next? opts) "src/cljs/chestnut/core_om_next.cljs"
                (rum? opts) "src/cljs/chestnut/core_rum.cljs"
                (vanilla? opts) "src/cljs/chestnut/core_vanilla.cljs"
                (re-frame? opts) "src/cljs/chestnut/core_re_frame.cljs")
              data)])))

(defn re-frame-render
  [resource-path data]
  (-> (str "leiningen/new/re_frame/" resource-path)
      io/resource
      slurp-resource
      (render-text data)))

(defn re-frame-files [data]
  (for [f ["config" "db" "subs" "events" "views"]]
    [(str "src/cljs/{{sanitized}}/" f ".cljs")
     (re-frame-render (str "src/cljs/" f ".cljs") data)]))

(defn do-pop-poll [version flags]
  (try
    (http/http-post! "https://lambdaisland.com/chestnut-poll" {:version version, :flags (str/join " " flags)})
    (catch Throwable e)))

(defn print-help! []
  (main/info "Chestnut: friendly Clojure/ClojureScript application template")
  (main/info "")
  (main/info "Usage: lein new chestnut <app-name> <options>")
  (main/info " e.g.: lein new chestnut dating-for-hamsters +http-kit +re-frame +code-of-conduct")
  (main/info "")
  (doseq [opt all-options]
    (main/info (format "  +%-20s %s" opt (options-descriptions opt))))
  (main/info "")
  (main/info "Please consult the Chestnut README (https://github.com/plexus/chestnut) as well as the README in the generated project."))

(defn chestnut [name & opts]
  (when (some #{"+help" "--help"} (cons name opts))
    (print-help!)
    (System/exit 0))

  (let [dash-opts (map (partial str "--") all-options)
        plus-opts (map (partial str "+") all-options)]
    (doseq [opt opts]
      (when (not (some #{opt} (concat dash-opts plus-opts)))
        (print-help!)
        (apply main/abort "Unrecognized option:" opt ". Should be one of" plus-opts))))

  (main/info "Generating fresh Chestnut project.")
  (main/info "README.md contains instructions to get you started.")

  (let [data (template-data name opts)
        files (cond-> (format-files-args data opts)
                (re-frame? opts) (concat (re-frame-files data)))]
    (apply ->files data files))

  (when-not (no-poll? opts)
    (do-pop-poll (chestnut-version) (map #(str/replace % #"^[-\+]+" "") opts)))

  (git-init name)
  (let [repo (load-repo name)]
    (git-add repo ".")
    (git-commit repo (str "lein new chestnut " name " " (str/join " " opts)))))
